peps/pep-0735.rst

```diff
@@ -0,0 +1,476 @@
+which has, as its value, a list of requirements specifiers (defined below).
+These keys must match the following regular expression:
+``[a-z0-9][a-z0-9-]*[a-z0-9]``. Meaning that they must be all lower-case
+alphanumerics, with ``-`` allowed only in the middle, and at least
+two characters long. These requirements are chosen so that the normalization
+rules used for pypi package names do not apply, as the names are already
+normalized.
+
+Requirements specifiers will use a definition which extends PEP 508. Therefore,
+it is first necessary to define a string format, a "PEP 735 Dependency". These
+are defined as strings in one of the three following formats:
+
+* a PEP 508 specification, e.g., ``numpy>1``
+* a comma-delimited list of Dependency Group names, enclosed in square
+  brackets, e.g., ``[test,docs]`` to refer to other Dependency Groups named
+  ``test`` and ``docs``
+* a single dot, ``.``, which refers to the current project as a package
+  (similar to ``pip install .``)
+* a single dot, followed by square brackets enclosing a list of extra names,
+  which refers to the current project as a package including some extras, e.g.,
+  ``.[mysql]`` to refer to the current package with its ``mysql`` extra
+
+Requirement specifiers can now be defined as one of the following:
+
+* A string, which is a valid PEP 735 Dependency Specifier. e.g., ``numpy>1``
+* An object, which has exactly one key, ``spec``, which is a valid PEP 735
+  Dependency Spec. e.g., ``{spec = "numpy>1"}``
+
+Any additional keys in a requirement specifier object are reserved for future
+use.
+
+The following is an example of a ``pyproject.toml`` section which uses this to
+define four dependency groups: ``test``, ``docs``, ``typing``, and
+``typing-test``:
+
+.. code:: toml
+
+    [requirements.packages]
+    test = ["pytest", "coverage", "."]
+    docs = ["sphinx", "sphinx-rtd-theme"]
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+Tools which support Dependency Groups MUST support both string and object
+representations of requirements.
+
+When unrecognized keys are encountered in requirement specifiers or the
+``requirements`` table, tools MUST NOT fail. They MAY emit warnings.
+
+When installing the current package (``.``) from a dependency group, tools
+SHOULD prefer editable installs over non-editable installs. They MAY provide
+users with options to configure this behavior.
```

Comment:
```suggestion
When installing the current package (``.``) from a dependency group, tools
SHOULD prefer editable installs over non-editable installs. They MAY provide
users with options to configure this behavior.
```

I personally don't like the "SHOULD" part of this. For instance, I don't use a `src/` layout typically, so an editable install is an unnecessary for me and all I really need is the dependencies installed. See https://github.com/pypa/pip/issues/11440 for a whole discussion about wanting to avoid the whole editable install step.

Reply:
Paul Moore raised this on Discourse as well. I hadn't realized that wanting to install `project.dependencies` without an install of the package was a use-case of importance.

I need to think more about how to specify this all. It seems to me that the things you might want to include are
- the current package (including `project.dependencies`)
- `project.dependencies` (sans the current package)
- any extra from `project.optional-dependencies` (IMO, *always* implicitly including `project.dependencies`)

If you _do_ use a `src/` layout, it makes a big difference if you include the current package, so I wouldn't want to omit it in all cases.

Reply:
The current version defines `only_deps` as a key in the object spec, taking direct inspiration from the pip thread, and `editable` as an explicit flag. The result is for the most part easier to reason about, and it forced me to formalize a lot of previously unspecified behaviors.

When `editable` is absent, this draft states explicitly that tools may choose their default/preference.

Between these new fields, I think these concerns are mostly addressed. At least, it's different enough from the earlier draft that I think we should have a new thread to discuss it further.